### PR TITLE
prevents drag and drop import in same window

### LINKED ISSUE #592: Disable drag and drop import within same window
It is currently too easy to inadvertently drag a link or image within the same wiki, accidentally creating a new tiddler.

We should detect and suppress  attempts to import from the same window.

### PR DESCRIPTION
**Pull request by tobibeer**
fixes #592

Introduces and preserves **_origin** field when using drag-and-drop...

* bad idea?
* discard?
* different field name?

demo http://592.tiddlyspot.com

### CODE REVIEW

File: core/modules/widgets/dropzone.js [hunk 1]
```diff
@@ -27,14 +27,14 @@ DropZoneWidget.prototype = new Widget();
 Render this widget into the DOM
 */
 DropZoneWidget.prototype.render = function(parent,nextSibling) {
-	var self = this;
```
Comment (Jermolene) [inline core/modules/widgets/dropzone.js]: It looks like you're making a coding style update at the same time as making functional modifications. In general, I think it's better to make them separate requests, as having two things going on can make it harder to understand and verify a request. But I must stress I'm not against cleaning up of the wrinkles in the code; it's tricky work, though, as it's hard to systematically test such changes.

Comment (tobibeer) [inline core/modules/widgets/dropzone.js]: It wasn't used, so I removed it. Is that a statement you always expect to be present?

File: core/modules/widgets/link.js [hunk 2]
```diff
@@ -153,7 +153,7 @@ LinkWidget.prototype.handleDragStartEvent = function(event) {
 			}
 			// Then the data
 			dataTransfer.clearData();
-			var jsonData = this.wiki.getTiddlerAsJson(this.to),
+			var jsonData = this.wiki.getTiddlerAsJson(this.to,true),
```
Comment (Jermolene) [inline core/modules/widgets/link.js]: I don't like the 'exporting' parameter for getTiddlerAsJson. Wouldn't it be clearer to add `jsonData["_origin"] = window.location.href` here?

Comment (tobibeer) [inline core/modules/widgets/link.js]: So, we first generate the json and then inject the **_origin** field... can do that, I was just thinking that this **_origin** thing may just end up being used in exactly this manner elsewhere.

**Comment by Jermolene** [comment]
Interesting start, thanks @tobibeer 

I'm not keen on the way that we're adding a temporary field `_origin` to tiddlers in transit. I'd prefer to update the mechanism to cope with a richer format that allows additional metadata to be passed along with the payload. Not sure how we can do that with backwards compatibility...

File: core/modules/widgets/dropzone.js (hunk 1 above)
Comment (Jermolene) [inline core/modules/widgets/dropzone.js]: No, not at all, it's just that there are plenty of those extraneous declarations. If we care about them, we should probably be applying lint and doing them in bulk. It's just to try to keep the pull requests simple to understand, and it is a general observation: in many specific cases where a function is being heavily modified these kind of modifications don't get in the way. In this particular case it just makes the fairly simple mods look more complex at first sight.

File: core/modules/widgets/link.js (hunk 2 above)
Comment (Jermolene) [inline core/modules/widgets/link.js]: Apologies, I just realised jsonData is a string.

Comment (Jermolene) [inline core/modules/widgets/link.js]: Perhaps a more general modification to getTiddlerAsJson would be to pass an optional hashmap of fields that are superimposed over the fields of the specified tiddler (including `undefined` causing the field to be omitted from the JSON).

**Comment by tobibeer** [comment]
Not sure if this is the right context but that origin field is every much of need, imho.

**Comment by tobibeer** [comment]
Shall I close this for now?

**Comment by Jermolene** [comment]
> but that origin field is very much of need, imho.

Maybe so, but it goes against the grain to be modifying the tiddlers that are imported. It makes it harder to later determine whether it is the same tiddler.

If we consider the `_origin` field as a component of a hypothetical individual tiddler subscribe/sync: we'd probably want to hold the data about the synchronisation subscription independently of the data item being subscribed.

Hmmm thinking about it a bit more, and considering @RichShumaker's recent discussion, perhaps a wiki should be able to declare fields and tiddlers that are automatically added to all tiddlers that are dragged to other wikis. So I could have a "copyright" field that is automatically affixed to tiddlers as they are exported, but that I don't have to have cluttering up the wiki while I'm editing it.

**Comment by Jermolene** [comment]
> Shall I close this for now?

I'm happy to continue figuring out a workable solution, this is something I'd love to fix.

**Comment by tobibeer** [comment]
> Maybe so, but it goes against the grain to be modifying the tiddlers that are imported. It makes it harder to later determine whether it is the same tiddler.

Actually, that may not be true. It makes it much clearer what we're supposedly importing. What is missing here is that I am bluntly setting an **_origin** field without checking if it may already be present first... and also, upon droping, checking that origin field rather than `window.location.href`... so "same tiddler" does not mean "from the same origin as the document" but rather "from the same origin as an existing tiddler in the document"... which makes for the truly interesting part.

At that moment, though... we do want the import, as we're likely upgrading some external component.

**Comment by Jermolene** [comment]
Hi @tobibeer thinking about it, I do think it might be a lot less disruptive to fix this with a global variable. In the drag handler set `$tw.dragging` to `true` while the drag is going on, and in the dragover handler ignore the drag if `$tw.dragging` is `true`.

**Comment by tobibeer** [comment]
Ok, so I'll close this particular pull request. How to unset `$tw.dragging` when leaving the window?

**Comment by Jermolene** [comment]
> How to unset $tw.dragging when leaving the window?

See: `LinkWidget.handleDragEndEvent()` in widgets/link.js

**Comment by tobibeer** [comment]
Ok, I hae pushed the alternative method that ignores any **_origin** for the time being

**Pull request merged**